haskell-src/Concordium/Types/KonsensusV1.hs

```diff
@@ -0,0 +1,134 @@
+            [ "block" .= qcBlock,
+              "round" .= qcRound,
+              "epoch" .= qcEpoch,
+              "aggregateSignature" .= qcAggregateSignature,
+              "signatories" .= qcSignatories
+            ]
+
+-- |The finalizers (identified by their 'BakerId's)
+-- that signed off for in the @frRound@.
+data FinalizerRound = FinalizerRound
+    { -- |The round.
+      frRound :: !Round,
+      -- |The finalizers who signed off in the round.
+      frFinalizers :: ![BakerId]
+    }
+    deriving (Eq, Show)
+
+instance ToJSON FinalizerRound where
+    toJSON FinalizerRound{..} =
+        object
+            [ "round" .= frRound,
+              "finalizers" .= frFinalizers
+            ]
+
+newtype TimeoutCertificateSignature = TimeoutCertificateSignature Bls.Signature
+    deriving (Eq, Show, ToJSON, FromJSON, Serialize) via Bls.Signature
+
+data TimeoutCertificate = TimeoutCertificate
+    { -- |The round that has timed-out.
+      tcRound :: !Round,
+      -- |The minimum epoch for which we include signatures.
+      tcMinEpoch :: !Epoch,
+      -- |The rounds for which finalizers have their best QCs in the epoch 'tcMinEpoch'.
+      tcFinalizerQCRoundsFirstEpoch :: ![FinalizerRound],
+      -- |The rounds for which finalizers have their best QCs in the epoch @tcMinEpoch + 1@.
+      tcFinalizerQCRoundsSecondEpoch :: ![FinalizerRound],
+      -- |Aggregate of the finalizers' 'TimeoutSignature's on the round and QC round.
+      tcAggregateSignature :: !TimeoutCertificateSignature
+    }
+    deriving (Eq, Show)
+
+instance ToJSON TimeoutCertificate where
+    toJSON TimeoutCertificate{..} =
+        object
+            [ "round" .= tcRound,
+              "minEpoch" .= tcMinEpoch,
+              "finalizerQCRoundsFirstEpoch" .= tcFinalizerQCRoundsFirstEpoch,
+              "finalizerQCRoundsSecondEpoch" .= tcFinalizerQCRoundsSecondEpoch,
+              "aggregateSignature" .= tcAggregateSignature
+            ]
+
+newtype SuccessorProof = SuccessorProof Hash.Hash
+    deriving (Eq, Show, ToJSON, FromJSON, Serialize) via Hash.Hash
+
+-- |The epoch finalization entry is the proof required in order to
+-- advance to a new epoch.
+data EpochFinalizationEntry = EpochFinalizationEntry
+    { -- |The qc that is finalized by the successor qc.
+      efeFinalizedQC :: !QuorumCertificate,
+      -- |The qc that finalizes @efeFinalizedQC@.
+      efeSuccessorQC :: !QuorumCertificate,
+      -- |A proof that the successor qc points to a block
+      -- which is an immediate successor of the block that
+      -- @efeFinalizedQC@ points to.
+      efeSuccessorProof :: !SuccessorProof
+    }
+    deriving (Eq, Show)
+
+instance ToJSON EpochFinalizationEntry where
+    toJSON EpochFinalizationEntry{..} =
+        object
+            [ "finalizedQC" .= efeFinalizedQC,
+              "successorQC" .= efeSuccessorQC,
+              "successorProof" .= efeSuccessorProof
+            ]
+
+-- |Block certificates for a block in 'ConsensusV1'.
```

Comment:
Is it `C` or `K` then?

Reply:
See above.

Reply:
As above, recommend "consensus version 1".